src/Symfony/Contracts/HttpClient/HttpClientTrait.php

```diff
@@ -0,0 +1,381 @@
+
+    /**
+     * Validates and normalizes method, URL and options, and merges them with defaults.
+     *
+     * When provided, the $defaultOptions MUST have been already validated and normalized.
+     *
+     * @throws \InvalidArgumentException&TransportExceptionInterface When a not-supported option is found
+     */
+    private static function prepareRequest(?string $method, ?string $url, array $options, array $defaultOptions = [], bool $allowExtraOptions = false): array
+    {
+        // Validate method
+        if (null !== $method && \strlen($method) !== strspn($method, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ')) {
+            throw self::createInvalidArgumentException(sprintf('Invalid HTTP method "%s", only uppercase letters are accepted.', $method), __FILE__, __LINE__);
+        }
+
+        // Option query is never inherited from defaults
+        $options['query'] = $options['query'] ?? [];
+
+        // Normalize body
+        if (isset($options['json'])) {
+            $options['body'] = self::jsonEncode($options['json']);
+            $defaultOptions['headers']['content-type'] = $defaultOptions['headers']['content-type'] ?? ['application/json'];
+        }
+
+        if (($body = $options['body'] ?? '') instanceof \Traversable) {
+            $body = function () use ($body) { yield from $body; };
+        }
+
+        if (\is_array($body)) {
+            $options['body'] = http_build_query($body, '', '&', PHP_QUERY_RFC1738);
+        } elseif ($body instanceof \Closure) {
+            $r = new \ReflectionFunction($body);
+            $body = $r->getClosure();
+
+            if ($r->isGenerator()) {
+                $body = $body(16372);
+                $body = function () use ($body) {
+                    $chunk = $body->valid() ? $body->current() : '';
+                    $body->next();
+
+                    return $chunk;
+                };
+            }
+
+            $options['body'] = $body;
+        } elseif (!\is_string($body) && !\is_array(@stream_get_meta_data($body))) {
+            throw self::createInvalidArgumentException(sprintf('Option "body" must be string, stream resource, iterable or callable, %s given.', \is_resource($body) ? get_resource_type($body) : \gettype($body)), __FILE__, __LINE__);
+        }
+
+        // Normalize headers
+        $rawHeaders = $headers = [];
+
+        foreach ($options['headers'] ?? [] as $name => $values) {
+            if (\is_int($name)) {
+                [$name, $values] = explode(':', $values, 2);
+                $values = [ltrim($values)];
+            } elseif (!\is_iterable($values)) {
+                $values = (array) $values;
+            }
+
+            $headers[$name = strtolower($name)] = [];
+
+            foreach ($values as $value) {
+                $headers[$name][] = $value;
+            }
+        }
+
+        // Merge default headers
+        $headers += $defaultOptions['headers'] ?? [];
```

Comment:
shouldnt we merge before normalizing headers

Reply:
> When provided, the $defaultOptions MUST have been already validated and normalized.

So, nope.

Reply:
I changed this to relax the constraint.